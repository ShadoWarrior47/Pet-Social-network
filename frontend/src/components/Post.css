/* General Post container styling */
.Post {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    border: 1px solid #ddd;
    border-radius: 8px;
    padding: 10px;
    margin: 10px 0;
    background-color: #f9f9f9;
    box-shadow: 0 2px 5px rgba(0, 0, 0, 0.1);
    transition: box-shadow 0.3s ease;
}

.Post:hover {
    box-shadow: 0 4px 10px rgba(0, 0, 0, 0.2);
}

.Post h2{
    margin: 0px auto;
}
.Post h3 {
    font-size: 1.5rem;
    margin: 8px 0px;
    color: #333;
}

.Post p {
    margin: 4px 0;
    font-size: 1rem;
    color: #555;
}

.Post img{
    width: 550px;
    margin: 2px auto;
}

.like-button-outline {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 3px 8px;
    font-size: 11px;
    font-weight: bold;
    margin-left: 15px;
    color: #ff4757;
    background-color: transparent;
    border: 2px solid #ff4757;
    border-radius: 50px;
    cursor: pointer;
    transition: background-color 0.3s ease, color 0.3s ease, transform 0.2s ease;
  }
  
  .like-button-outline:hover {
    background-color: #ff4757;
    color: #ffffff;
  }
  
  .like-button-outline:active {
    transform: scale(0.95);
  }
  

  .like-button-pulse {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    padding: 3px 8px;
    font-size: 11px;
    font-weight: bold;
    margin-left: 15px;
    color: #ffffff;
    background-color: #ff4757;
    border: none;
    border-radius: 50px;
    cursor: pointer;
    animation: pulse 1.5s infinite;
  }
  
  @keyframes pulse {
    0%, 100% {
      box-shadow: 0 0 0 0 rgba(255, 71, 87, 0.4);
    }
    50% {
      box-shadow: 0 0 20px 20px rgba(255, 71, 87, 0);
    }
  }
  