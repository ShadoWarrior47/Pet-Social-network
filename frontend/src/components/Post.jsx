
import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import PetApi from "../PetApi";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faComment as fasComment } from '@fortawesome/free-solid-svg-icons';
import "./Post.css";

function Post({ post, currentUser, onLike, onDelete, isMyPost = false }) {
  const isOwner = currentUser && currentUser.id === post.ownerId;

  const [comments, setComments] = useState([]);
  const [likeCount, setLikeCount] = useState(0);
  const [liked, setLiked] = useState(post.isLiked || false);

  // Fetch likes count and comments on component mount
  useEffect(() => {
    async function fetchPostData() {
      if (currentUser) {
        try {
          // Fetch likes count
          const count = await PetApi.getLikesCount(post.id);
          setLikeCount(count);

          // Fetch comments
          const fetchedComments = await PetApi.getComments(post.id);
          setComments(fetchedComments);
        } catch (err) {
          console.error("Error fetching post data:", err);
        }
      }
    }
    fetchPostData();
  }, [post.id, currentUser]);

  // Fetch like status on component mount
  useEffect(() => {
    async function fetchLikeStatus() {
      if (currentUser) {
        try {
          const isLiked = await PetApi.isPostLiked(post.id);
          setLiked(isLiked);
        } catch (err) {
          console.error("Error checking like status:", err);
        }
      }
    }
    fetchLikeStatus();
  }, [post.id, currentUser]);

  // Handle like/dislike action
  async function handleLike(postId) {
    if (!currentUser || !currentUser.id) {
      console.error("Error: User not authenticated or missing ID");
      return;
    }
    try {
      if (liked) {
        await PetApi.removeLike(postId); // Call the API to remove the like
        setLikeCount((prev) => prev - 1);
      } else {
        await PetApi.addLike(postId, { ownerId: currentUser.id }); // Call the API to add the like
        setLikeCount((prev) => prev + 1);
      }
      setLiked(!liked); // Toggle the liked state
    } catch (err) {
      console.error("Error liking/disliking post:", err);
    }
  }

  return (
    <div className="Post">
      <h2>
        <Link to={`/posts/${post.id}`}>{post.content}</Link>
      </h2>
      {post.imageUrl && <img src={post.imageUrl} alt="Post" />}
      {/* <p class = "tweet info">{post.ownerId} - {post.createdAt}</p> */}

      <div className="Post-actions">
        {currentUser && (
          <>
            <button onClick={() => handleLike(post.id)} className={liked ? "like-button-pulse" : "like-button-outline"}>
              {liked ? '❤️' : '♡'} Like
              ({likeCount})
            </button>

            <button className="comment-button">
              <FontAwesomeIcon icon={fasComment} />
              <Link to={`/posts/${post.id}`}>Comment</Link>
            </button>

          </>
        )}

        {isMyPost && (
          <>
            <button>
              <Link to={`/posts/edit/${post.id}`}>Edit</Link>
            </button>
            <button onClick={() => onDelete(post.id)}>Delete</button>
          </>
        )}
      </div>

      {currentUser && (
        <div className="Post-comments">
          <h4>Comments:</h4>
          {comments.slice(-2).map(comment => (
            <div key={comment.id} className="Post-comment">
              <strong>{comment.username}</strong>: {comment.content}
            </div>
          ))}
          <Link to={`/posts/${post.id}`}>View all comments</Link>
        </div>
      )}
    </div>
  );

}
export default Post;