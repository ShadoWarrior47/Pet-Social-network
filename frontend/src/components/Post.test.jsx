import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Post from "./Post";
import '@testing-library/jest-dom';

it("renders without crashing", function () {
    render(
        <MemoryRouter>
            <Post post={{ id: 5, content: "Test Post"}} />
        </MemoryRouter>
    );
});

it("matches snapshot", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <Post post={{ id: 5, content: "Test Post" }} />
        </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

it("displays pet details correctly", function () {
    const { getByText } = render(
        <MemoryRouter>
            <Post post={{ id: 5, content: "Test Post"}} />
        </MemoryRouter>
    );

    expect(getByText("Test Post")).toBeInTheDocument();
    // expect(getByText("This is a test company.")).toBeInTheDocument();
});
