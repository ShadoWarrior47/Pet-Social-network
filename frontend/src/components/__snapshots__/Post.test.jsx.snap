// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Post"
  >
    <h2>
      <a
        data-discover="true"
        href="/posts/5"
      >
        Test Post
      </a>
    </h2>
    <div
      class="Post-actions"
    />
  </div>
</DocumentFragment>
`;
